File: day_1/day1.py
def Day1_part1() :
    f = open('day1.txt', 'r')
    for line in f :
        directions = line.split(", ")

    directions[-1] = directions[-1].split('\n')[0]
    print directions

    actual_direction = "N"
    x = 0
    y = 0
    positions=[]
    positions.append([x,y])

    for d in directions :
        #print d[0]
        if d[0] == "R" and actual_direction == "N" :
            x += int(d[1:])
            actual_direction = "E"
            Day1_part2(positions,x,y)
            continue

        if d[0] == "L" and actual_direction == "N" :
            x -= int(d[1:])
            actual_direction = "W"
            Day1_part2(positions,x,y)
            continue

        if d[0] == "R" and actual_direction == "S" :
            x -= int(d[1:])
            actual_direction = "W"
            Day1_part2(positions,x,y)
            continue

        if d[0] == "L" and actual_direction == "S" :
            x += int(d[1:])
            actual_direction = "E"
            Day1_part2(positions,x,y)
            continue

        if d[0] == "R" and actual_direction == "E" :
            y -= int(d[1:])
            actual_direction = "S"
            Day1_part2(positions,x,y)
            continue

        if d[0] == "L" and actual_direction == "E" :
            y += int(d[1:])
            actual_direction = "N"
            Day1_part2(positions,x,y)
            continue

        if d[0] == "R" and actual_direction == "W" :
            y += int(d[1:])
            actual_direction = "N"
            Day1_part2(positions,x,y)
            continue

        if d[0] == "L" and actual_direction == "W" :
            y -= int(d[1:])
            actual_direction = "S"
            Day1_part2(positions,x,y)
            continue

    print abs(x)+abs(y)































def Day1_part2() :
        mapp = [ [ 0 for y in range( 500 ) ] for x in range( 500 ) ]
        f = open('day1.txt', 'r')
        for line in f :
            directions = line.split(", ")

        directions[-1] = directions[-1].split('\n')[0]
        print directions

        actual_direction = "N"
        x = 0
        y = 0
        positions=[]
        positions.append([x,y])

        for d in directions :
            #print d[0]
            if d[0] == "R" and actual_direction == "N" :
                for i in range(x, x+int(d[1:])+1) :
                    mapp[i][y] += 1
                    if mapp[i][y] == 2 :
                        print abs(i)+abs(y)
                x += int(d[1:])
                actual_direction = "E"
                continue

            if d[0] == "L" and actual_direction == "N" :
                for i in range(x, x-int(d[1:])-1,-1) :
                    mapp[i][y] += 1
                    if mapp[i][y] == 2 :
                        print abs(i)+abs(y)
                x -= int(d[1:])
                actual_direction = "W"
                continue

            if d[0] == "R" and actual_direction == "S" :
                for i in range(x, x-int(d[1:])-1,-1) :
                    mapp[i][y] += 1
                    if mapp[i][y] == 2 :
                        print abs(i)+abs(y)
                x -= int(d[1:])
                actual_direction = "W"
                continue

            if d[0] == "L" and actual_direction == "S" :
                for i in range(x, x+int(d[1:])+1) :
                    mapp[i][y] += 1
                    if mapp[i][y] == 2 :
                        print abs(i)+abs(y)
                x += int(d[1:])
                actual_direction = "E"
                continue

            if d[0] == "R" and actual_direction == "E" :
                for i in range(y, y-int(d[1:])-1,-1) :
                    mapp[x][i] += 1
                    if mapp[x][i] == 2 :
                        print abs(x)+abs(i)
                y -= int(d[1:])
                actual_direction = "S"
                continue

            if d[0] == "L" and actual_direction == "E" :
                for i in range(y, y-int(d[1:])+1) :
                    mapp[x][i] += 1
                    if mapp[x][i] == 2 :
                        print abs(x)+abs(i)
                y += int(d[1:])
                actual_direction = "N"
                continue

            if d[0] == "R" and actual_direction == "W" :
                for i in range(y, y-int(d[1:])+1) :
                    mapp[x][i] += 1
                    if mapp[x][i] == 2 :
                        print abs(x)+abs(i)
                y += int(d[1:])
                actual_direction = "N"
                continue

            if d[0] == "L" and actual_direction == "W" :
                for i in range(y, y-int(d[1:])-1,-1) :
                    mapp[x][i] += 1
                    if mapp[x][i] == 2 :
                        print abs(x)+abs(i)
                y -= int(d[1:])
                actual_direction = "S"
                continue

        #print abs(x)+abs(y)

Day1_part2()

File: day_1/day1.java
import java.io.*;
import java.util.*;

public class day1 {

  public static void main(String[] args) throws FileNotFoundException, IOException {
    String input = "";

    // Read the input.
    try(BufferedReader br = new BufferedReader(new FileReader("day1.txt"))) {
      String line = br.readLine();

      while (line != null) {
        input += line;
        line = br.readLine();
      }
    }

    // Split the input string to a List of Strings.
    List<String> inputList = Arrays.asList(input.split(", "));


    Part_1(inputList);
    System.out.println("............................");
    Part_2(inputList);


  }




  public static void Part_1(List<String> inputList){
    int x = 0;
    int y = 0;

    Direction compass = Direction.NORTH;
    for (int i=0; i<inputList.size(); i++){

      if (compass == Direction.NORTH && inputList.get(i).charAt(0) == 'R'){
        compass = Direction.EAST;
        x += Integer.parseInt(inputList.get(i).substring(1));
        continue;
      }

      if (compass == Direction.NORTH && inputList.get(i).charAt(0) == 'L'){
        compass = Direction.WEST;
        x -= Integer.parseInt(inputList.get(i).substring(1));
        continue;
      }

      if (compass == Direction.SOUTH && inputList.get(i).charAt(0) == 'R'){
        compass = Direction.WEST;
        x -= Integer.parseInt(inputList.get(i).substring(1));
        continue;
      }

      if (compass == Direction.SOUTH && inputList.get(i).charAt(0) == 'L'){
        compass = Direction.EAST;
        x += Integer.parseInt(inputList.get(i).substring(1));
        continue;
      }

      if (compass == Direction.EAST && inputList.get(i).charAt(0) == 'R'){
        compass = Direction.SOUTH;
        y -= Integer.parseInt(inputList.get(i).substring(1));
        continue;
      }

      if (compass == Direction.EAST && inputList.get(i).charAt(0) == 'L'){
        compass = Direction.NORTH;
        y+= Integer.parseInt(inputList.get(i).substring(1));
        continue;
      }

      if (compass == Direction.WEST && inputList.get(i).charAt(0) == 'R'){
        compass = Direction.NORTH;
        y += Integer.parseInt(inputList.get(i).substring(1));
        continue;
      }

      if (compass == Direction.WEST && inputList.get(i).charAt(0) == 'L'){
        compass = Direction.SOUTH;
        y-= Integer.parseInt(inputList.get(i).substring(1));
        continue;
      }
    }
    System.out.println("Santa's route length: " + (Math.abs(x) + Math.abs(y)));
  }















  public static void Part_2(List<String> inputList){
    int x = 500;
    int y = 500;

    boolean[][] map = new boolean[1000][1000];

    Direction compass = Direction.NORTH;
    for (int i=0; i<inputList.size(); i++){
      int path = Integer.parseInt(inputList.get(i).substring(1));

      if (compass == Direction.NORTH && inputList.get(i).charAt(0) == 'R'){
        compass = Direction.EAST;
        for (int j=0; j<path; j++){
          if (!map[x][y+j])
            map[x][y+j] = true;
          else {
            System.out.println("First location visited twice: " + x + " " + (y+j));
            return;
          }
        }
        y += path;
        continue;
      }

      if (compass == Direction.NORTH && inputList.get(i).charAt(0) == 'L'){
        compass = Direction.WEST;
        for (int j=0; j<path; j++){
          if (!map[x][y-j])
            map[x][y-j] = true;
          else {
            System.out.println("First location visited twice: " + x + " " + (y-j));
            return;
          }
        }
        y -= path;
        continue;
      }

      if (compass == Direction.SOUTH && inputList.get(i).charAt(0) == 'R'){
        compass = Direction.WEST;
        for (int j=0; j<path; j++){
          if (!map[x][y-j])
            map[x][y-j] = true;
          else {
            System.out.println("First location visited twice: " + x + " " + (y-j));
            return;
          }
        }
        y -= path;
        continue;
      }

      if (compass == Direction.SOUTH && inputList.get(i).charAt(0) == 'L'){
        compass = Direction.EAST;
        for (int j=0; j<path; j++){
          if (!map[x][y+j])
            map[x][y+j] = true;
          else {
            System.out.println("First location visited twice: " + x + " " + (y+j));
            return;
          }
        }
        y += path;
        continue;
      }

      if (compass == Direction.EAST && inputList.get(i).charAt(0) == 'R'){
        compass = Direction.SOUTH;
        for (int j=0; j<path; j++){
          if (!map[x+j][y])
            map[x+j][y] = true;
          else {
            System.out.println("First location visited twice: " + (x+j) + " " + y);
            return;
          }
        }
        x += path;
        continue;
      }

      if (compass == Direction.EAST && inputList.get(i).charAt(0) == 'L'){
        compass = Direction.NORTH;
        for (int j=0; j<path; j++){
          if (!map[x-j][y])
            map[x-j][y] = true;
          else {
            System.out.println("First location visited twice: " + (x-j) + " " + y);
            return;
          }
        }
        x -= path;
        continue;
      }

      if (compass == Direction.WEST && inputList.get(i).charAt(0) == 'R'){
        compass = Direction.NORTH;
        for (int j=0; j<path; j++){
          if (!map[x-j][y])
            map[x-j][y] = true;
          else {
            System.out.println("First location visited twice: " + (x-j) + " " + y);
            return;
          }
        }
        x -= path;
        continue;
      }

      if (compass == Direction.WEST && inputList.get(i).charAt(0) == 'L'){
        compass = Direction.SOUTH;
        for (int j=0; j<path; j++){
          if (!map[x+j][y])
            map[x+j][y] = true;
          else {
            System.out.println("First location visited twice: " + (x+j) + " " + y);
            return;
          }
        }
        x += path;
        continue;
      }

    }
  }



}


enum Direction {
  NORTH, SOUTH, EAST, WEST
}
